### PR TITLE
Feature/ibft round change manager

### PR DESCRIPTION
**Pull request by Errorific**
## PR description
Manager for dealing with IBFT round change messages inside a round

### CODE REVIEW

File: consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    }
+
+    public void addMessage(final SignedData<RoundChangePayload> msg) {
+      if (!actioned) {
+        receivedMessages.put(msg.getSender(), msg);
+      }
+    }
+
+    public boolean roundChangeReady() {
+      return receivedMessages.size() >= quorumSize && !actioned;
```
Comment (rain-on) [inline consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java]: now wondering if "actioned" is being used correctly - i.e. why can't someone make multiple RoundChangeCertificates?

File: consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+ * proposer for said round) - a newRound message is sent, and a new round should be started by the
+ * controlling class.
+ */
+public class RoundChangeManager {
+
+  public static class RoundChangeStatus {
+    private final int quorumSize;
+
+    @VisibleForTesting
+    final Map<Address, SignedData<RoundChangePayload>> receivedMessages = Maps.newHashMap();
```
Comment (rain-on) [inline consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java]: Assuming this is map such that if 2 messages from 1 validator are received, the previous one is overwritten? Probably worth a comment.

File: consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        actioned = true;
+        return new RoundChangeCertificate(receivedMessages.values());
+      } else {
+        throw new IllegalStateException("Unable to create RoundChangeCertificate at this time.");
+      }
+    }
+  }
+
+  private static final Logger LOG = LogManager.getLogger();
+  @VisibleForTesting
```
Comment (rain-on) [inline consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java]: concerned about "VisibleForTesting" on internal data, both here and above.

File: consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    final RoundChangeStatus roundChangeStatus = roundChangeCache.get(msgTargetRound);
+    roundChangeStatus.addMessage(msg);
+    if (roundChangeStatus.roundChangeReady()) {
+      return Optional.of(roundChangeStatus.createRoundChangeCertificate());
+    }
+    return Optional.empty();
+  }
+
+  /* Called when a given round has completed/timed-out such that msgs cached can be removed */
+  public void discardCompletedRound(final ConsensusRoundIdentifier completedRoundIdentifier) {
```
Comment (rain-on) [inline consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java]: Checkout using Collection.removeif - probably makes this a one-liner.
roundChangeCache.entrySet().removeIf(e -> e.getKey)

File: consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    // validate the received msg
+    final ConsensusRoundIdentifier msgTargetRound = msg.getPayload().getRoundChangeIdentifier();
+    final RoundChangeMessageValidator messageValidator =
+        new RoundChangeMessageValidator(
+            messageValidityFactory, validators, quorumSize, sequenceNumber);
+    if (!messageValidator.validateMessage(msg)) {
+      LOG.info("RoundChange message was invalid.");
+      return Optional.empty();
+    }
+    if (!roundChangeCache.containsKey(msgTargetRound)) {
```
Comment (rain-on) [inline consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java]: Look at breaking some of this out to a separate function. Or maybe breaking it in half.

File: consensus/ibft/src/test/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManagerTest.java [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+  public void rejectsInvalidRoundChangeMessage() {
+    SignedData<RoundChangePayload> roundChangeData = makeRoundChangeMessage(nonValidatorKey, ri1);
+    assertThat(manager.appendRoundChangeMessage(roundChangeData)).isEqualTo(Optional.empty());
+    assertThat(manager.roundChangeCache.get(ri1)).isNull();
+  }
+
+  @Test
+  public void acceptsValidRoundChangeMessage() {
+    SignedData<RoundChangePayload> roundChangeData = makeRoundChangeMessage(proposerKey, ri2);
+    assertThat(manager.appendRoundChangeMessage(roundChangeData)).isEqualTo(Optional.empty());
```
Comment (rain-on) [inline consensus/ibft/src/test/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManagerTest.java]: can change "isEqualTo(Optional.empty())" into ".isEmpty()", here and below.

File: consensus/ibft/src/test/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManagerTest.java [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+    SignedData<RoundChangePayload> roundChangeDataValidator1 =
+        makeRoundChangeMessage(validator1Key, ri2);
+    assertThat(manager.appendRoundChangeMessage(roundChangeDataProposer))
+        .isEqualTo(Optional.empty());
+    assertThat(manager.appendRoundChangeMessage(roundChangeDataValidator1))
+        .satisfies(Optional::isPresent);
+  }
+
+  @Test
+  public void doesntReadyWithDifferentRounds() {
```
Comment (rain-on) [inline consensus/ibft/src/test/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManagerTest.java]: Is this named appropriately, or is it missing a word?

File: consensus/ibft/src/test/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManagerTest.java [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+            mock(MutableBlockchain.class), mock(WorldStateArchive.class), mock(IbftContext.class));
+
+    @SuppressWarnings("unchecked")
+    BlockHeaderValidator<IbftContext> headerValidator =
+        (BlockHeaderValidator<IbftContext>) mock(BlockHeaderValidator.class);
+    BlockHeader parentHeader = mock(BlockHeader.class);
+
+    Map<ConsensusRoundIdentifier, MessageValidator> messageValidators = Maps.newHashMap();
+
+    messageValidators.put(
```
Comment (rain-on) [inline consensus/ibft/src/test/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManagerTest.java]: Personally, I think these can be mocked to produce required outputs.

File: consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java (hunk 3 above)
Comment (Errorific) [reply consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java]: I'm open to suggestions of how to test without it.

File: consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java (hunk 1 above)
Comment (Errorific) [reply consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java]: What was the purpose in including it in the original code?

File: consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+
+  public Optional<RoundChangeCertificate> appendRoundChangeMessage(
+      final SignedData<RoundChangePayload> msg) {
+    // validate the received msg
+    final ConsensusRoundIdentifier msgTargetRound = msg.getPayload().getRoundChangeIdentifier();
+    final RoundChangeMessageValidator messageValidator =
+        new RoundChangeMessageValidator(
+            messageValidityFactory, validators, quorumSize, sequenceNumber);
+    if (!messageValidator.validateMessage(msg)) {
+      LOG.info("RoundChange message was invalid.");
```
Comment (jframe) [inline consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java]: I think the WARN level makes more sense than info as this is potentially an problem something should look at

File: consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java (hunk 4 above)
Comment (Errorific) [reply consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java]: done

File: consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java (hunk 2 above)
Comment (Errorific) [reply consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java]: done

File: consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java (hunk 5 above)
Comment (Errorific) [reply consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java]: done

File: consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+      return Optional.empty();
+    }
+    if (!roundChangeCache.containsKey(msgTargetRound)) {
```
Comment (jframe) [inline consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java]: Can use putIfAbsent to simplify this i.e. something like roundChangeCache.putIfAbsent(msgTargetRound, new RoundChangeStatus(quorumSize));

File: consensus/ibft/src/test/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManagerTest.java (hunk 7 above)
Comment (Errorific) [reply consensus/ibft/src/test/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManagerTest.java]: clarified

File: consensus/ibft/src/test/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManagerTest.java [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+  }
+
+  @Test
+  public void doesntAcceptDuplicateValidRoundChangeMessage() {
+    SignedData<RoundChangePayload> roundChangeData = makeRoundChangeMessage(proposerKey, ri2);
+    assertThat(manager.appendRoundChangeMessage(roundChangeData)).isEmpty();
+    assertThat(manager.appendRoundChangeMessage(roundChangeData)).isEmpty();
+    assertThat(manager.roundChangeCache.get(ri2).receivedMessages.size()).isEqualTo(1);
+  }
+
```
Comment (jframe) [inline consensus/ibft/src/test/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManagerTest.java]: Do we need a similar test to check that we don't accept further round change message after reaching the threshold?

Comment (Errorific) [reply consensus/ibft/src/test/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManagerTest.java]: If we stick with the actioned behaviour yes. Depending on the outcome to @rain-on's earlier comment I'll add a test or adjust.

File: consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java (hunk 10 above)
Comment (Errorific) [reply consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManager.java]: or .computeIfAbsent(). Done

File: consensus/ibft/src/test/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManagerTest.java (hunk 11 above)
Comment (Errorific) [reply consensus/ibft/src/test/java/tech/pegasys/pantheon/consensus/ibft/statemachine/RoundChangeManagerTest.java]: Test is added

**Pull request merged**